sdk/inc/azure/core/az_json.h

```diff
@@ -522,14 +562,47 @@ typedef struct
  * will use the default options (i.e. #az_json_reader_options_default()).
  *
  * @return An #az_result value indicating the result of the operation:
- *         - #AZ_OK if the az_json_reader is initialized successfully
- *         - #AZ_ERROR_EOF if the provided json buffer is empty
+ *         - #AZ_OK if the #az_json_reader is initialized successfully
+ *
+ * @remarks The provided json buffer must not be empty, as that is invalid JSON.
+ *
+ * @remarks An instance of #az_json_reader must not outlive the lifetime of the JSON payload within
+ * the \p json_buffer.
  */
 AZ_NODISCARD az_result az_json_reader_init(
     az_json_reader* json_reader,
     az_span json_buffer,
     az_json_reader_options const* options);
 
+/**
+ * @brief Initializes an #az_json_reader to read the JSON payload contained within the provided
+ * set of discontiguous buffers.
+ *
+ * @param[out] json_reader A pointer to an #az_json_reader instance to initialize.
+ * @param[in] json_buffers An array of non-contiguous byte buffers, as spans, containing the JSON
+ * text to read.
+ * @param[in] number_of_buffers The number of buffer segments provided, i.e. the length of the \p
+ * json_buffers array.
+ * @param[in] options __[nullable]__ A reference to an #az_json_reader_options
+ * structure which defines custom behavior of the #az_json_reader. If `NULL` is passed, the reader
+ * will use the default options (i.e. #az_json_reader_options_default()).
+ *
+ * @return An #az_result value indicating the result of the operation:
+ *         - #AZ_OK if the #az_json_reader is initialized successfully
+ *
+ * @remarks The provided array of json buffers must not be empty, as that is invalid JSON, and
+ * therefore \p number_of_buffers must also be greater than 0. The array must also not contain any
+ * empty span segments.
+ *
+ * @remarks An instance of #az_json_reader must not outlive the lifetime of the JSON payload within
+ * the \p json_buffers.
+ */
+AZ_NODISCARD az_result az_json_reader_chunked_init(
+    az_json_reader* json_reader,
+    az_span json_buffers[],
```

Comment:
Should we terminate with AZ_SPAN_NULL instead of number_of_buffers? It would be more consistent with our other APIs.

Reply:
I thought about doing that, but the detection logic for "end of buffer" would be something like this:

```C
if (az_span_size(*json_buffers) == 0)
{
   // done, no more  data
}
```

We don't support empty segments today, but if we were to allow those in the future, such a check would break. So, having an explicit `number_of_buffers` enables us to change the checks in the future and enable empty buffers.

I want to limit and avoid cases (ideally 0) where we differentiate between an empty span and a null span sentinel.

Part of the rationale is similar to having char* and length APIs rather than relying on the char* ending with a null terminator to find the length.

What do you think?

In the interest of time, I am going to file an issue to track this.

Reply:
https://github.com/Azure/azure-sdk-for-c/issues/1027

Reply:
there is a `AZ_LOG_END_OF_LIST` for adding classification list end in azure/core/az_log.h

Maybe add something like `AZ_JSON_END_OF_BUFFER` to end the array instead of an AZ_SPAN_NULL?

Reply:
Interesting idea. To me, adding `AZ_LOG_END_OF_LIST` as another enum entry makes sense. Adding another public sentinel `az_span` value just for this particular case seems unnecessary.